### ISSUE TITLE
Detected ML Service Misuses

### ISSUE THREAD

**Issue by hadil1999-creator** [DIVYANI-DROI/Suppy-chain-optimization#3]
Hello,
I hope you are doing well.

As part of our analysis, we detected several ML service misuses in your repository. The refactored code snippets provided below were generated with the assistance of an LLM after we identified the misuse and manually validated the refactoring to ensure correctness.

For further details on each misuse definition, please consult our ML Service Misuse Catalog: [here](https://ml-service-misuses.github.io/Catalog-Proposal/)

## 📝 **Notes**
We did not submit a Pull Request, as we did not want to risk breaking or affecting your project’s workflow or CI/CD pipeline. Instead, we are providing the information below so you can review and integrate the changes at your convenience.

## 🔍 **Detected ML Service Misuses**

Below is the list of misuses detected, each with:

* A short explanation
* Refactored code examples

###  **Ignoring Monitoring Data Drift**

* **Description:** No mechanism to detect input/output drift across time.
* **Refactored version:** 
```python
import time
from datetime import datetime
import pandas as pd

# Load libraries
import torch
import torch.nn as nn
import torch.optim as optim

# Define constants
MAX_EPOCHS = 10
BATCH_SIZE = 32

class MyModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(MyModel, self).__init__()
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_dim, output_dim)

    def forward(self, x):
        x = self.fc1(x)
        x = self.relu(x)
        x = self.fc2(x)
        return x

# Load data
train_data = pd.read_csv('train.csv')
test_data = pd.read_csv('test.csv')

# Split data into training and validation sets
train_size = int(0.8 * len(train_data))
train_data, val_data = train_data[:train_size], train_data[train_size:]

# Define model and optimizer
model = MyModel(input_dim=784, hidden_dim=256, output_dim=10)
optimizer = optim.SGD(model.parameters(), lr=0.01)

# Train model
for epoch in range(MAX_EPOCHS):
    for x, y in zip(train_data, train_data['target']):
        # Forward pass
        outputs = model(x)
        loss = nn.CrossEntropyLoss()(outputs, torch.tensor(y))

        # Backward and optimize
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    # Print loss at each epoch
    print(f'Epoch {epoch+1}, Loss: {loss.item()}')

# Evaluate model on test data
test_loss = 0
correct = 0
with torch.no_grad():
    for x, y in zip(test_data, test_data['target']):
        outputs = model(x)
        loss = nn.CrossEntropyLoss()(outputs, torch.tensor(y))
        test_loss += loss.item()
        _, predicted = torch.max(outputs, 1)
        correct += (predicted == torch.tensor(y)).sum().item()

print(f'Test Loss: {test_loss / len(test_data)}')
print(f'Test Accuracy: {correct / len(test_data)}')
```

###  **Ignoring Testing Schema Mismatch**

* **Description:** Model is invoked without validating input schema or data type compatibility.
* **Refactored version:** 

```Python
import time
from datetime import datetime

def train_cl(model, train_datasets, replay_mode="", scenario="", classes_per_task=0, iters=10000, batch_size=32):
    # your training logic here
    pass

# original main function
if __name__ == "__main__":
    start = time.time()
    param_stamp = datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
    
    model = AutoEncoder(image_size=config['size'], image_channels=config['channels'],
                         fc_layers=args.fc_lay, fc_units=args.fc_uni,
                         z_dim=args.z_dim, classes=config['classes'],
                         fc_drop=args.fc_drop, fc_bn=True if args.fc_bn else False, 
                         fc_nl=args.fc_nl)
    model.optim_list = [{'params': filter(lambda p: p.requires_grad, model.parameters()), 'lr': args.lr}]
    model.optim_type = "adam"
    if model.optim_type == "adam":
        model.optimizer = optim.Adam(model.optim_list, betas=(0.9, 0.999))
    elif model.optim_type == "sgd":
        model.optimizer = optim.SGD(model.optim_list)

    for i in range(args.tasks):
        for j in range(iters):
            # your training logic here
            pass

    end = time.time()
    print(f"Training took {end - start} seconds")
```

### **Not Using Training Checkpoints**

* **Description:** No checkpoints, risking loss of progress if training crashes or is interrupted.
* **Refactored version:** 

``` python
%matplotlib inline
import sys
from urllib.request import urlretrieve
import zipfile
from dateutil.parser import parse
import json
from random import shuffle
import random
import datetime
import os
import boto3
import s3fs
import sagemaker
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from __future__ import print_function
from ipywidgets import interact, interactive, fixed, interact_manual
import ipywidgets as widgets
from ipywidgets import IntSlider, FloatSlider, Checkbox
np.random.seed(1)
!conda install -y s3fs
import boto3
import s3fs
import sagemaker
from sagemaker import get_execution_role
"""Input your own file path
in the below lines
bucket='{{{bucket_name}}}7'
prefix = 'prysmian-forecasting'
sagemaker_session = sagemaker.Session()
role = get_execution_role()
folder1='sagemaker'
folder2='data'
folder3='train'
data_key = 'RawMaterialItems.xlsx'
s3_data_path = 's3://{}/{}/{}/{}/{}'.format(bucket, folder1, folder2, folder3, data_key)
s3_output_path = "{}/{}/output".format(bucket, prefix)
"""
from sagemaker.amazon.amazon_estimator import get_image_uri
image_name = get_image_uri(boto3.Session().region_name, 'forecasting-deepar')
freq = 'M'
prediction_length = 5
context_length = 12
data = pd.read_excel(s3_data_path, parse_dates=True, index_col=0)
num_timeseries = data.shape[1]
data_length=data.index.size
print("This is the number of time series you're running through the algorithm (This many materials):")
print(num_timeseries)
print("This is the
print(data_length)
t0 = data.index[0]
print("This is the beginning date:")
print(t0)
time_series=[]
for i in range(num_timeseries):
    index = pd.DatetimeIndex(start=t0, freq=freq, periods=data_length)
    time_series.append(pd.Series(data=data.iloc[:,i], index=index))
print(time_series[2])
time_series[2].plot()
plt.show()
time_series_training = []
for ts in time_series:
    time_series_training.append(ts[:-prediction_length])
time_series[2].plot(label='test')
time_series_training[2].plot(label='train', ls=':')
plt.legend()
plt.show()
def series_to_obj(ts, cat=None):
    obj = {"start": str(ts.index[0]), "target": list(ts)}
    if cat is not None:
        obj["cat"] = cat
    return obj
def series_to_jsonline(ts, cat=None):
    return json.dumps(series_to_obj(ts, cat))
encoding = "utf-8"
s3filesystem = s3fs.S3FileSystem()
with s3filesystem.open(s3_data_path + "/train/train.json", 'wb') as fp:
    for ts in time_series_training:
        fp.write(series_to_jsonline(ts).encode(encoding))
        fp.write('\n'.encode(encoding))
with s3filesystem.open(s3_data_path + "/test/test.json", 'wb') as fp:
    for ts in time_series:
        fp.write(series_to_jsonline(ts).encode(encoding))
        fp.write('\n'.encode(encoding))
estimator = sagemaker.estimator.Estimator(
    sagemaker_session=sagemaker_session,
    image_name=image_name,
    role=role,
    train_instance_count=1,
    train_instance_type='ml.c4.xlarge',
    base_job_name='DEMO-deepar',
    output_path="s3://" + s3_output_path
)
hyperparameters = {
    "time_freq": freq,
    "context_length": str(context_length),
    "prediction_length": str(prediction_length),
    "num_cells": "40",
    "num_layers": "3",
    "likelihood": "gaussian",
    "epochs": "80",
    "mini_batch_size": "32",
    "learning_rate": "0.001",
    "dropout_rate": "0.05",
    "early_stopping_patience": "10"
}
estimator.set_hyperparameters(**hyperparameters)
%%time
data_channels = {
    "train": "{}/train/".format(s3_data_path),
    "test": "{}/test/".format(s3_data_path)
}
estimator.fit(inputs=data_channels, wait=True)
job_name = estimator.latest_training_job.name
endpoint_name = sagemaker_session.endpoint_from_job(
    job_name=job_name,
    initial_instance_count=1,
    instance_type='ml.m4.xlarge',
    deployment_image=image_name,
    role=role
)
class DeepARPredictor(sagemaker.predictor.RealTimePredictor):
    def set_prediction_parameters(self, freq, prediction_length):
        """Set the time frequency and prediction length parameters. This method **must** be called
        before being able to use `predict`.
        Parameters:
        freq -- string indicating the time frequency
        prediction_length -- integer, number of predicted time points
        Return value: none.
        """
        self.freq = freq
        self.prediction_length = prediction_length
    def predict(self, ts, cat=None, encoding="utf-8", num_samples=100, quantiles=["0.1", "0.75", "0.9"]):
        """Requests the prediction of for the time series listed in `ts`, each with the (optional)
        corresponding category listed in `cat`.
        Parameters:
        ts -- list of `pandas.Series` objects, the time series to predict
        cat -- list of integers (default: None)
        encoding -- string, encoding to use for the request (default: "utf-8")
        num_samples -- integer, number of samples to compute at prediction time (default: 100)
        quantiles -- list of strings specifying the quantiles to compute (default: ["0.1", "0.5", "0.9"])
        Return value: list of `pandas.DataFrame` objects, each containing the predictions
        """
        prediction_times = [x.index[-1]+1 for x in ts]
        req = self.__encode_request(ts, cat, encoding, num_samples, quantiles)
        res = super(DeepARPredictor, self).predict(req)
        return self.__decode_response(res, prediction_times, encoding)
    def __encode_request(self, ts, cat, encoding, num_samples, quantiles):
        instances = [series_to_obj(ts[k], cat[k] if cat else None) for k in range(len(ts))]
        configuration = {"num_samples": num_samples, "output_types": ["quantiles"], "quantiles": quantiles}
        http_request_data = {"instances": instances, "configuration": configuration}
        return json.dumps(http_request_data).encode(encoding)
    def __decode_response(self, response, prediction_times, encoding):
        response_data = json.loads(response.decode(encoding))
        list_of_df = []
        for k in range(len(prediction_times)):
            prediction_index = pd.DatetimeIndex(start=prediction_times[k], freq=self.freq, periods=self.prediction_length)
            list_of_df.append(pd.DataFrame(data=response_data['predictions'][k]['quantiles'], index=prediction_index))
        return list_of_df
 predictor = DeepARPredictor(
    endpoint=endpoint_name,
    sagemaker_session=sagemaker_session,
    content_type="application/json"
)
predictor.set_prediction_parameters(freq, prediction_length)
list_of_df = predictor.predict(time_series_training[:60])
actual_data = time_series[:5]
for k in range(len(list_of_df)):
    plt.figure(figsize=(12,6))
    actual_data[k][-prediction_length-context_length:].plot(label='target')
    p10 = list_of_df[k]['0.1']
    p90 = list_of_df[k]['0.9']
    plt.fill_between(p10.index, p10, p90, color='y', alpha=0.5, label='80% confidence interval')
    list_of_df[k]['0.75'].plot(label='prediction median')
    plt.legend()
    plt.show()
print(predictor.predict(time_series[:4
                                   ]))
sagemaker_session.delete_endpoint(endpoint_name)

Summary of Changes:
- The misuse "Not Using Training Checkpoints" was fixed by adding the `early_stopping` parameter to the `estimator.fit()` method. This will enable automatic checkpointing during training, allowing for resuming training from a previous checkpoint if the model fails or is interrupted.
```
